tests/unit/deferred.py

```diff
@@ -0,0 +1,222 @@
+    return model.Nic(
+        model.Node(
+            label='node-99',
+            obm=MockObm(
+                type="http://schema.massopencloud.org/haas/v0/obm/mock",
+                host="ipmihost",
+                user="root",
+                password="tapeworm")),
+        name,
+        '00:11:22:33:44:55')
+
+
+@pytest.fixture()
+def network():
+    project = model.Project('anvil-nextgen')
+    return model.Network(project, [project], True, '102', 'hammernet')
+
+pytestmark = pytest.mark.usefixtures('configure',
+                                     'fail_on_log_warnings')
+
+
+def test_apply_networking(switch, network, fresh_database):
+    '''Test to validate apply_networking commits actions incrementally
+
+    This test verifies that the apply_networking() function in hil/deferred.py
+    incrementally commits actions, which ensures that any error on an action
+    will not require a complete rerun of the prior actions (e.g. if an error
+    is thrown on the 3rd action, the 1st and 2nd action will have already been
+    committed)
+
+    The test also verifies that if a new networking action fails, then the
+    old networking actions in the queue were commited.
+    '''
+    nic = []
+    action = []
+    # initialize 3 nics and networking actions
+    for i in range(0, 3):
+        interface = 'gi1/0/%d' % (i)
+        nic.append(new_nic(str(i)))
+        nic[i].port = model.Port(label=interface, switch=switch)
+        action.append(model.NetworkingAction(nic=nic[i], new_network=network,
+                                             channel='vlan/native',
+                                             type='modify_port'))
+
+    # this makes the last action invalid for the test switch which will raise
+    # an exception.
+    action[2] = model.NetworkingAction(nic=nic[2],
+                                       new_network=None,
+                                       channel='',
+                                       type='revert_port')
+    db.session.add(action[0])
+    db.session.add(action[1])
+    db.session.add(action[2])
+    db.session.commit()
+    try:
+        deferred.apply_networking()
+    except SwitchError:
+        # close the session opened by `apply_networking` when `handle_actions`
+        # fails; without this the tests would just stall (when using postgres)
+        db.session.close()
+
+        local_db = new_db()
+
+        pending_action = local_db.session \
+            .query(model.NetworkingAction) \
+            .order_by(model.NetworkingAction.id).first()
+        current_count = local_db.session \
+            .query(model.NetworkingAction).count()
+
+        local_db.session.delete(pending_action)
+        local_db.session.commit()
+        local_db.session.close()
+
+        # test that there's only pending action in the queue and it is of type
+        # revert_port
+        assert current_count == 1
+        assert pending_action.type == 'revert_port'
```

Comment:
The way this is organized now, if something goes wrong and a `SwitchError` doesn't get raised at all, this check will just get skipped. I'd suggest:

```python
with pytest.raises(SwitchError):
     deferred.apply_networking()

# Everything that's currently in the except block.
```

... which will fail the test if the exception is not raised.